docs/csharp/tutorials/string-interpolation.md

```diff
@@ -26,6 +26,8 @@ As the example shows, you include an expression in an interpolated string by enc
 
 At compile time, an interpolated string is typically transformed into a <xref:System.String.Format%2A?displayProperty=nameWithType> method call. That makes all the capabilities of the [string composite formatting](../../standard/base-types/composite-formatting.md) feature available to you to use with interpolated strings as well.
 
+If all expressions are strings, and none of them have alignment or format specifiers, an interpolated string will be transformed into a more efficient string concatenation <xref:System.String.Concat%2A?displayProperty=nameWithType> at compile time.
```

Comment:
I wonder how much of the implementation details should be exposed in the docs. How does knowing it help me to *use* interpolated strings? Note, the paragraph above uses the "typically" word, so it already allows the implementation to use other method instead of `string.Format`. The main motivation behind mentioning of the `string.Format` method was to make a link to the composite formatting feature. The message is "hey, folks! you can easily convert your existing `string.Format` calls into the interpolated string; not a big deal, the core formatting features are the same". Looking at it now, I think that paragraph can be rephrased the following way:
> Interpolated strings support all the capabilities of the [string composite formatting](../../standard/base-types/composite-formatting.md) feature. That makes them a more readable alternative to the use of the `<xref:System.String.Format%2A?displayProperty=nameWithType>` method.

That way, the main idea is the same and implementation details are even more covered. And then, most of the next sections show how to use those "capabilities".

@OlehPetrenko @BillWagner thoughts?
@OlehPetrenko anyway, thanks for the PR; that way I've learnt about that optimization 👍

Reply:
Thanks for your review and good comment.

I thought about the "typically" word that you mentioned, but as for me this change in the Roslyn is too important to hide it under the "typically" word.
A lot of people think that string interpolation is just a syntactic sugar for a String.Format, but after the commit it is not a true, and it is not a easy to find right commit in the Roslyn to understand what happens.

> How does knowing it help me to use interpolated strings?

Just for for example, if you need to write a performance sensitive code and you want to use integer value as part of interpolated string. You can write 5.ToString() instead of 5 to avoid boxing and call String.Contac that is more efficient than String.Format. 

My point is, this paragraph can save time to understand why sometimes you see String.Concat instead of String.Format in the IL.

Reply:
@OlehPetrenko thanks for the additional context.

Do I understand it correctly, that knowledge (compilation to the `string.Concat` call) is important, because one (one, working on the performance sensitive code, at least) might decide to use string interpolation instead of manually crafting calls to the `string.Concat`? Then, I think that knowledge is more valuable to answer WHETHER one should use string interpolation, rather than HOW to use string interpolation. The tutorial is about *how*; it might be a little bit late to provide the implementation detail.

If that info is to be included into the docs, then I think it should be mentioned in the main article about the feature. If I want to decide if to use the feature, I would start with the main article about it, not tutorial. The main article is:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated

Maybe, the separate section just before the "Additional resources" section should be added to discuss `string.Format`/`string.Concat`.

Also, the language version, from which this optimization is available, should be mentioned. If that knowledge is important to make any decisions from it, one also needs to know WHEN this information is relevant/correct. I guess, the implemented optimization is not for C#6 and C# 7.x, but available starting with C# 8.0? I mean, if I write code using C# 7.0, I still must write the calls to `string.Concat` myself?

Reply:
@pkulikov yes, you are right about performance sensivice code, otherwise this optimization would not be added to the Roslyn, I guess. Also, this knowledge can be useful for everyone who wants to know what happens under the hood.

I've decided to add this paragraph here, because here is the place where mentioned how it works at compile time. A lot of information exists at the StackOverFlow and other resources where it says that interpolation is the same thing with String.Format. It was true up to a #26612 issue, but now it may confuse. The official documentation is the best way to help people.

I think, the paragraph that I suggested is just the easiest way to provide people with this new behavior, but separate section it is also a good idea, but maybe it could be redundant.

Anyway, if my paragraph is not be approved, it would be great if existed paragraph about String.Format would be rephrased.

It is not depends on language version. Only depends on Roslyn's version. I've checked on 2.9.0.63208 and this oprimization works.

Reply:
@OlehPetrenko 

> I've decided to add this paragraph here, because here is the place where mentioned how it works at compile time.

>Anyway, if my paragraph is not be approved, it would be great if existed paragraph about String.Format would be rephrased.

That makes sense. I think the existing paragraph should be rephrased in any case. I've made an attempt in my first comment, but maybe it's better to address it in another PR.

Implementation details, both about `string.Format` and `string.Concat`, fit better another article. At least, the tutorial is not the best place for them, I feel.

>It doesn't depend on language version. Only depends on Roslyn's version. I've checked on 2.9.0.63208 and this oprimization works.

Thanks for the check and clarification. It's then about the tools installed (.NET Core SDK or Visual Studio).

By the way, have you noticed the follow-up PR dotnet/roslyn#27411 that limits the optimization scope?

Reply:
Yes, I've seen it. 

Thanks for your comments.